### PR TITLE
Xdlops gemm output optimization

### PR DESCRIPTION
**Pull request by zjing14**
- Add a generic 10-dimensional tensor description for writing output of xdlops gemm from register to global memory in a single copy
- Improve fp16 performance up to 10%, no performance drop for fp32 after retune
- Based on PR #323
- Hold due to performance regression

### CODE REVIEW

**Comment by atamazov** [comment]
[off-topic] @zjing14 I recommend using drafts instead of "WIP" in the PR title:
image

**Comment by asleepzzz** [comment]
merge into develop?

**Comment by asroy** [comment]
@zjing14 Does this PR relies on #353? I merge this only PR with #347, and test these configs. I'm seeing some regression.

I tested 2 cases in below with #347 only @1189Mhz. Please check if you can get the same performance with this PR.

```
bin/MIOpenDriver convfp16 -F 1 -n 128 -g 1 -k 2048 -c 2048 -H  35 -W   36 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -V 0 -w 1 -t 1 -i 1
MIOpen(OpenCL): Info [FindConvFwdAlgorithm] miopenConvolutionFwdAlgoImplicitGEMM        11.6692 0
MIOpen(OpenCL): Info [FindConvFwdAlgorithm] FW Chosen Algorithm: ConvHipImplicitGemmForwardV4R4Xdlops , 0, 11.6692
MIOpen(OpenCL): Info [ConvolutionForward] algo = 5, workspace = 0
Wall-clock Time Forward Conv. Elapsed: 12.2866 ms, Auxiliary API calls: 4801.42 ms (GWSS: 23.334)
MIOpen Forward Conv. Algorithm: 5, Solution: 64/ConvHipImplicitGemmForwardV4R4Xdlops
GPU Kernel Time Forward Conv. Elapsed: 11.683955 ms (average)
stats: name, n, c, ho, wo, x, y, k, flopCnt, bytesRead, bytesWritten, GFLOPs, GB/s, timeMs
stats: fwd-conv1x1u1, 128, 2048, 35, 36, 1, 1, 2048,  1352914698240, 668991488, 660602880, 115793, 114, 11.683955
```
```
bin/MIOpenDriver convfp16 -F 1 -n   1 -g 1 -k 4096 -c 4096 -H   1 -W 3840 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -V 0 -w 1 -t 1 -i 1
MIOpen(OpenCL): Info [FindConvFwdAlgorithm] miopenConvolutionFwdAlgoImplicitGEMM        1.14928 0
MIOpen(OpenCL): Info [FindConvFwdAlgorithm] FW Chosen Algorithm: ConvHipImplicitGemmForwardV4R4Xdlops , 0, 1.14928
MIOpen(OpenCL): Info [ConvolutionForward] algo = 5, workspace = 0
Wall-clock Time Forward Conv. Elapsed: 1.6176 ms, Auxiliary API calls: 4593.7 ms (GWSS: 22.5532)
MIOpen Forward Conv. Algorithm: 5, Solution: 64/ConvHipImplicitGemmForwardV4R4Xdlops
GPU Kernel Time Forward Conv. Elapsed: 1.145755 ms (average)
stats: name, n, c, ho, wo, x, y, k, flopCnt, bytesRead, bytesWritten, GFLOPs, GB/s, timeMs
stats: fwd-conv1x1u1, 1, 4096, 1, 3840, 1, 1, 4096,  128849018880, 65011712, 31457280, 112458, 84, 1.145755
```

**Comment by zjing14** [comment]
> @zjing14 Does this PR relies on #353? I merge this only PR with #347, and test these configs. I'm seeing some regression.
> 
> I tested 2 cases in below with #347 only @1189Mhz. Please check if you can get the same performance with this PR.
> 
> ```
> bin/MIOpenDriver convfp16 -F 1 -n 128 -g 1 -k 2048 -c 2048 -H  35 -W   36 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -V 0 -w 1 -t 1 -i 1
> MIOpen(OpenCL): Info [FindConvFwdAlgorithm] miopenConvolutionFwdAlgoImplicitGEMM        11.6692 0
> MIOpen(OpenCL): Info [FindConvFwdAlgorithm] FW Chosen Algorithm: ConvHipImplicitGemmForwardV4R4Xdlops , 0, 11.6692
> MIOpen(OpenCL): Info [ConvolutionForward] algo = 5, workspace = 0
> Wall-clock Time Forward Conv. Elapsed: 12.2866 ms, Auxiliary API calls: 4801.42 ms (GWSS: 23.334)
> MIOpen Forward Conv. Algorithm: 5, Solution: 64/ConvHipImplicitGemmForwardV4R4Xdlops
> GPU Kernel Time Forward Conv. Elapsed: 11.683955 ms (average)
> stats: name, n, c, ho, wo, x, y, k, flopCnt, bytesRead, bytesWritten, GFLOPs, GB/s, timeMs
> stats: fwd-conv1x1u1, 128, 2048, 35, 36, 1, 1, 2048,  1352914698240, 668991488, 660602880, 115793, 114, 11.683955
> ```
> 
> ```
> bin/MIOpenDriver convfp16 -F 1 -n   1 -g 1 -k 4096 -c 4096 -H   1 -W 3840 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -V 0 -w 1 -t 1 -i 1
> MIOpen(OpenCL): Info [FindConvFwdAlgorithm] miopenConvolutionFwdAlgoImplicitGEMM        1.14928 0
> MIOpen(OpenCL): Info [FindConvFwdAlgorithm] FW Chosen Algorithm: ConvHipImplicitGemmForwardV4R4Xdlops , 0, 1.14928
> MIOpen(OpenCL): Info [ConvolutionForward] algo = 5, workspace = 0
> Wall-clock Time Forward Conv. Elapsed: 1.6176 ms, Auxiliary API calls: 4593.7 ms (GWSS: 22.5532)
> MIOpen Forward Conv. Algorithm: 5, Solution: 64/ConvHipImplicitGemmForwardV4R4Xdlops
> GPU Kernel Time Forward Conv. Elapsed: 1.145755 ms (average)
> stats: name, n, c, ho, wo, x, y, k, flopCnt, bytesRead, bytesWritten, GFLOPs, GB/s, timeMs
> stats: fwd-conv1x1u1, 1, 4096, 1, 3840, 1, 1, 4096,  128849018880, 65011712, 31457280, 112458, 84, 1.145755
> ```

Tuned. Same performance with the first config

```
$ MIOPEN_FIND_ENFORCE=3 ~/igemm_run.sh 
stats: name, n, c, ho, wo, x, y, k, flopCnt, bytesRead, bytesWritten, GFLOPs, GB/s, timeMs
stats: fwd-conv1x1u1, 128, 2048, 35, 36, 1, 1, 2048,  1352914698240, 668991488, 660602880, 114106, 112, 11.856680
```

Regression with the second config.
```
MIOpenDriver convfp16 -F 1 -n 1 -g 1 -k 4096 -c 4096 -H 1 -W 3840 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -V 0 -w 1 -t 1 -i 1
stats: name, n, c, ho, wo, x, y, k, flopCnt, bytesRead, bytesWritten, GFLOPs, GB/s, timeMs
stats: fwd-conv1x1u1, 1, 4096, 1, 3840, 1, 1, 4096,  128849018880, 65011712, 31457280, 91930, 69, 1.401604
```

File: src/kernels/composable_kernel/include/tensor_operation/gridwise_gemm_xdlops_fp16_bfp16.hpp [hunk 1]
```diff
@@ -986,61 +985,63 @@ struct GridwiseBatchGemmXdlops_gkmkpack_gknkpack_gmn_v2
         {
             ///\todo inconsistent layout of xdlops and tensor
```
Comment (TejashShah) [inline src/kernels/composable_kernel/include/tensor_operation/gridwise_gemm_xdlops_fp16_bfp16.hpp:986]: Please remove /todo if it is no longer relevant.

**Pull request closed without merging**